### ISSUE TITLE
Feature: Xonsh equivalent to IPython whos

### ISSUE THREAD

**Issue by JohnLunzer** [scopatz/xonsh#936]
```python
SESSION_START_CONTEXT = locals()
def diffContext(args, stdin=None):
    from builtins import __xonsh_ctx__
    curr_ctx = {k:v for k,v in __xonsh_ctx__.items() if k is not '__name__'}
    diff_ctx_set = set(curr_ctx) ^ set(SESSION_START_CONTEXT)
    diff_ctx = {k:(curr_ctx[k], type(curr_ctx[k])) for k in diff_ctx_set}
    print(diff_ctx)
aliases['whox'] = diffContext
```
In xonsh this is even more important than in IPython because sometimes the user creates naming conflicts with system commands. Maybe this exist already but I couldn't find it. 

Could be pronounce "hooch" or "whoz" depending on how you translate the 'x'. 

Let me know what you think.

**Comment by scopatz** [comment]
Probably best not to pronounce it "hooch," but I like the idea! I wouldn't mind just calling it `whos`, since that would be what people expect. Feel free to put in a PR!

**Comment by scopatz** [comment]
PS I don't think this exists already...

**Comment by JohnLunzer** [comment]
You're right, ```whos``` is best. I can also add a ```who``` rather easily.

I plan on submitting a pull request for this. But I have two questions:

- IPython already has a lovely ```whos``` command which I have already heavily modified and cleaned of dependencies to make it suitable for xonsh. I'm not really familiar with licensing issues so I'm not sure what I need to do to inject the modified code into xonsh. 

- Where is the best place to add these commands? Should I add them directly to aliases.py or would they be better suited elsewhere?

**Comment by gforsyth** [comment]
Hey @JohnLunzer -- you'll need to include the IPython license and so it might be best if your `whos` fork was in its own file (to better delineate the code under said license).  It could be that `xoreutils` is a good place for this, although it isn't a "standard" bit of GNU core stuff.  

```
IPython is licensed under the terms of the Modified BSD License (also known as New or Revised or 3-Clause BSD), as follows:

    Copyright (c) 2008-2014, IPython Development Team
    Copyright (c) 2001-2007, Fernando Perez <fernando.perez@colorado.edu>
    Copyright (c) 2001, Janko Hauser <jhauser@zscout.de>
    Copyright (c) 2001, Nathaniel Gray <n8gray@caltech.edu>

All rights reserved.

Redistribution and use in source and binary forms, with or without modification, are permitted provided that the following conditions are met:

Redistributions of source code must retain the above copyright notice, this list of conditions and the following disclaimer.

Redistributions in binary form must reproduce the above copyright notice, this list of conditions and the following disclaimer in the documentation and/or other materials provided with the distribution.

Neither the name of the IPython Development Team nor the names of its contributors may be used to endorse or promote products derived from this software without specific prior written permission.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
```

**Comment by JohnLunzer** [comment]
thanks @gforsyth , I'll wait a bit in case anyone else wants to comment.

**Comment by denfromufa** [comment]
whox?

**Comment by scopatz** [comment]
Hi @JohnLunzer - Since IPython is BSD licensed. All that you need to include is the copyright notice, technically (not the whole license).  This could go in xoreutils, but as @gforsyth mentions, it really isn't a coreutil analog. I think it might make the most sense to just have a whos.py file, unless this is less than 20 lines, in which case aliases.py is fine.

**Comment by scopatz** [comment]
aliases.py is for small aliases and the Aliases class, while more involved functionality (xontrib, xonfig, etc) are in their own files.

**Comment by JohnLunzer** [comment]
@scopatz @gforsyth, I have a question. Right now the what I have `whos` implemented in my `.xonshrc` file I rely on the code:
```python
SESSION_START_CONTEXT = {k:v for k,v in locals().items()}
```
being at the very end of my `.xonshrc` file. This makes the most sense to me as it ensures that when I first start a session and type whos that I get the message `No new variables since start of session`.

Where is the best place in the code to place this so it executes after .xonshrc is processed and so I can access it from the `whos` command?

**Comment by scopatz** [comment]
@JohnLunzer, I think that `__xonsh_ctx__` is what you want instead.

**Comment by JohnLunzer** [comment]
I do use `__xonsh_ctx__` later but I can certainly use for the start context as well. 

But what I'm really interested in is when and where to save my `SESSION_START_CONTEXT` object/variable.

**Comment by scopatz** [comment]
Well, isn't `SESSION_START_CONTEXT` effectively the same as `__xonsh_ctx__`?  Also, if it as at the end of the xonshrc it will go into the `__xonsh_ctx__`, right? So it can be looked up from there if you need it.

**Comment by JohnLunzer** [comment]
If I'm not mistaken `__xonsh_ctx__` is the *current* state of xonsh's local namespace. The way I make who work is by saving the `__xonsh_ctx__` at the start of the xonsh session and then comparing it to the current `__xonsh_ctx__` at the invocation of `whos`.

I'm interested in when/where in xonsh's startup process to save the `__xonsh_ctx__` which represents xonsh's local namespace at the start of the session. Based on how I have it defined in my own `.xonshrc` file I know this must happen right after processing the `.xonshrc` file so as to not clutter it with variables set in `.xonshrc`.

**Comment by scopatz** [comment]
I guess I would have to see a PR, but I think that a whos command should show you what is in the xonshrc because it is part of the local execution context it would be strange not to show it.  I can understand the argument for hiding variable that start with an underscore.

**Comment by tgoodlet** [comment]
+1 on this! It's something I really miss in `xonsh` right now.

You guys might want to also consider that `IPython` also defines a `%reset` magic:
```python
In [1]: reset?
Docstring:
Resets the namespace by removing all names defined by the user, if
called without arguments, or by removing some types of objects, such
as everything currently in IPython's In[] and Out[] containers (see
the parameters for details).
```
This is very handy for cleaning. 
Obviously reuse of the name `reset` isn't ideal since it collides with the *nix utility.

**Comment by scopatz** [comment]
PRs very welcome!

**Comment by tgoodlet** [comment]
@scopatz as long as @JohnLunzer is cool with it I may take a stab at it :)

**Comment by scopatz** [comment]
> @scopatz as long as @JohnLunzer is cool with it I may take a stab at it :)

I think this has been sitting long enough that you should feel free!

> I'm finally getting around to "what do I need in xonsh now that I'm running it as my default shell".

Help is always welcome!

**Comment by JohnLunzer** [comment]
@tgoodlet, yeah go for it. Initially I rolled my own code but the ipython code is pretty solid so I recommend starting there. You just need to alter how the whos funtion accesses environmental variables in xonsh. Don't forget to include the ipython license snippet if you use their code.

**Comment by tgoodlet** [comment]
@scopatz so here's the relevant [`IPython` code](https://github.com/ipython/ipython/blob/master/IPython/core/magics/namespace.py#L242). 

I want to discuss this a little further. 
This is what I've hand rolled in my `xonshrc` to get a basic prototype going:

```python
                                                                                         
# namespace bookkeeping                                                                  
_new = set()                                                                             
                                                                                         
                                                                                         
def ctxdiff(now, start, remove):                                                         
    new = now.keys() - start.keys() - remove                                             
    return {k: now[k] for k in new}                                                      
                                                                                         
                                                                                         
def diff_context(args, stdin=None):                                                      
    diff = ctxdiff(__xonsh_ctx__, __xonsh_ctx__['_startctx'], {'_startctx', '__name__'}) 
    _new.update(diff.keys())                                                             
    for k, v in diff.items():                                                            
        print('{} {}: {}'.format(type(v), k, v))                                         
                                                                                         
                                                                                         
def clear(args, stdin=None):                                                             
    global _new                                                                          
    for k in _new:                                                                       
        __xonsh_ctx__.pop(k)                                                             
    _new = set()                                                                         
                                                                                         
                                                                                         
aliases['ns'] = diff_context                                                             
aliases['cns'] = clear                                                                   
_startctx  = __xonsh_ctx__.copy()                                                        
```
Yes I realize I changed the name. `ns` (namespace) seems cleaner to me as I never liked the name `whos` but I honestly don't care too much. To me `ns` and `cns` fit the shorthand `matplotlib` has been using with figure/axes management from the CLI. 

The `IPython` shell maintains the notion of a *user namespace* (a *hidden* one as well) something, afaict, `xonsh` does not do and which I'm hacking together ^. 

So here's my questions:
- how do we go about defining a *user namespace* proper with `__xonsh__ctx__`?
- what's included in this *user namespace* (should we include config file definitions or only state from the interactive session)?
- what do we call the pre-user namespace snapshot so that I don't have to keep describing what it *is* every time I am talking about it ;P (pre-interact ns?)

@JohnLunzer asked a reasonable question about *when* (*where* in code) the initial snapshot of the user's namespace should be taken so that our `whos` only reports new state introduced by the user. If I add it in `xonsh.aliases` where am I getting that initial `copy()` from? I would think setting global state in that module is quite ugly. One thought I had while starting to browse the `xonsh` sources was using an event hook like [`on_post_rc`](https://github.com/xonsh/xonsh/blob/master/xonsh/shell.py#L153). My only concern is how to pass the the *pre-interact ns* data to the alias sanely. I guess we could always store the *pre-interact ns* as an entry in `__xonsh_ctx__`? I wasn't sure if that would be dirtying it up.

PS I'm not sure it should display new state defined in `xonsh` config files btw. `IPython` doesn't so why should we?

**Comment by scopatz** [comment]
Thanks for the detailed write up @tgoodlet!

My opinion is that we should *not* have a separate user namespace, and that there should only be `__xonsh_ctx__`.  (Cascading namespace lookup can be achieved with `collections.MultiMap`, but it is slower - which is bad since this is the globals dict that is hit all of the time. Also it complicates the model.)

Rather, I think that we should simply store the names (keys of `__xonsh_ctx__`) as a set of str keys at the beginning. Then everytime `ns` or `cns` it would do the set removal like you wrote above. Simply put, there is no need to store a (deep)copy of the original context, but just the keys.

Both mechanisms have their trade offs. If you store just the keys and the user deletes an original key, it is gone for good. If you store a copy of the original context and you reset, modifications made the original state (such as changes to a xontrib variable) are wiped out. Personally, I come down on the side that user modifications should stay, and so I think of storing just the keys as the least evil option - the user did delete the variable after all.

I also think that `on_pre_cmdloop` is probably the right event for this particular use case.  All that the handle should do for this is store the context keys in some other variable, something like,

```python
START_NAMES = frozenset()

@events.on_pre_cmdloop
def store_start_names():
    global START_NAMES
    START_NAMES = frozenset(__xonsh_ctx__.keys())
```

**Comment by tgoodlet** [comment]
>Rather, I think that we should simply store the names (keys of __xonsh_ctx__) as a set of str keys at the beginning. 

@scopatz good exactly what I was thinking :)

My only last question is then should we store `START_NAMES` in one of the builtins maybe? `__xonsh_ctx__` or something else? I guess it can always be done later if something else needs it.

**Comment by astronouth7303** [comment]
> My only last question is then should we store START_NAMES in one of the builtins maybe? `__xonsh_ctx__` or something else? I guess it can always be done later if something else needs it.

You can just keep it in the module, especially if it's a xontrib. It's all Python. If you need to inspect it from the shell, `from xontrib.whos import start_names` would be fine, yes?

**Comment by scopatz** [comment]
Yeah, I agree with @astronouth7303. This is an implementation detail of ns and doesn't need to be in the builtins.

**Comment by astronouth7303** [comment]
What's the status of this?

**Comment by tgoodlet** [comment]
@astronouth7303 I've got something small working and have been meaning to come back to this.
It's unfortunately really far down on my list atm.
I'd appreciate keeping this open if you guys don't mind.

**Comment by laerus** [comment]
maybe the distinction of xonsh's namespaces could be more clear, 
user variables, xonsh builtins, aliases, envvars, OS commands aka $PATH,
all valid namespaces with different context.

pluggable namespaces seems like a nice feature for core xonsh, could help #2417

**Comment by astronouth7303** [comment]
Define pluggable namespaces? Almost all those namespaces are selected by syntax:

* Python globals (shell context/user variables, xonsh builtins)
* envvars
* commands (aliases, programs)

Which namespace is used depends entirely on the syntax and is selected by the parser.

**Comment by laerus** [comment]
the name lookup procedure is hard coded atm. 
something simple like `__xonsh_ctx__` being a list of the different namespaces would make it possible of changing the order.

**Comment by laerus** [comment]
**the implementation is not the simple part the spec/API can be though.

**Comment by astronouth7303** [comment]
yeah, the lookup is actually done by CPython, not by xonsh code.
